execution/scan/functions.go

```diff
@@ -0,0 +1,678 @@
+		}
+		return sample{
+			T: f.StepTime,
+			F: f.Samples[len(f.Samples)-1].F,
+		}
+	},
+	"present_over_time": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: 1,
+		}
+	},
+	"changes": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: changes(f.Samples),
+		}
+	},
+	"resets": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: resets(f.Samples),
+		}
+	},
+	"deriv": func(f functionArgs) sample {
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: deriv(f.Samples),
+		}
+	},
+	"irate": func(f functionArgs) sample {
+		f.Samples = filterFloatOnlySamples(f.Samples)
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		val, ok := instantValue(f.Samples, true)
+		if !ok {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: val,
+		}
+	},
+	"idelta": func(f functionArgs) sample {
+		f.Samples = filterFloatOnlySamples(f.Samples)
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		val, ok := instantValue(f.Samples, false)
+		if !ok {
+			return invalidSample
+		}
+		return sample{
+			T: f.StepTime,
+			F: val,
+		}
+	},
+	"rate": func(f functionArgs) sample {
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		v, h := extrapolatedRate(f.Samples, true, true, f.StepTime, f.SelectRange, f.Offset)
+		return sample{
+			T: f.StepTime,
+			F: v,
+			H: h,
+		}
+	},
+	"delta": func(f functionArgs) sample {
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		v, h := extrapolatedRate(f.Samples, false, false, f.StepTime, f.SelectRange, f.Offset)
+		return sample{
+			T: f.StepTime,
+			F: v,
+			H: h,
+		}
+	},
+	"increase": func(f functionArgs) sample {
+		if len(f.Samples) < 2 {
+			return invalidSample
+		}
+		v, h := extrapolatedRate(f.Samples, true, false, f.StepTime, f.SelectRange, f.Offset)
+		return sample{
+			T: f.StepTime,
+			F: v,
+			H: h,
+		}
+	},
+	"xrate": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
+		}
+		if f.MetricAppearedTs == nil {
+			panic("BUG: we got some Samples but metric still hasn't appeared")
+		}
+		v, h := extendedRate(f.Samples, true, true, f.StepTime, f.SelectRange, f.Offset, *f.MetricAppearedTs)
+		return sample{
+			T: f.StepTime,
+			F: v,
+			H: h,
+		}
+	},
+	"xdelta": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
+		}
+		if f.MetricAppearedTs == nil {
+			panic("BUG: we got some Samples but metric still hasn't appeared")
+		}
+		v, h := extendedRate(f.Samples, false, false, f.StepTime, f.SelectRange, f.Offset, *f.MetricAppearedTs)
+		return sample{
+			T: f.StepTime,
+			F: v,
+			H: h,
+		}
+	},
+	"xincrease": func(f functionArgs) sample {
+		if len(f.Samples) == 0 {
+			return invalidSample
```

Comment:
I wonder if we should remove `invalidSample` in favour of an `ok` return argument to indicate whether the function returns any value. The return args would then be `float64, *histogram.FloatHistogram, bool`. We also don't need the timestamp of the result sample since that comes from the step itself.

Reply:
that sounds much better! thanks ill do that

Reply:
done! it has some effect

on main:
```
RangeQuery/clamp-8                                           632.5m ± ∞ ¹    102.5m ± ∞ ¹   -83.79% (p=0.008 n=5)
```
vs PR
```
RangeQuery/clamp-8                                          631.62m ± ∞ ¹    85.65m ± ∞ ¹   -86.44% (p=0.008 n=5)
```